File: services/scala-pekko/src/main/scala/stt/protocol/Protocol.scala
package stt.protocol

import io.circe.derivation.{Configuration, ConfiguredEncoder}
import io.circe.{Decoder, DecodingFailure, Encoder, Json}

object Protocol:
  val FrameBytes: Int = 640
  val PartialType: String = "partial"
  val ErrorType: String = "error"

  val CloseProtocolError: Int = 1002
  val CloseUnsupportedData: Int = 1003
  val ReasonNeedStart: String = "first message must be start"
  val ReasonTextAfterStart: String = "expected binary PCM frames after start"
  val ReasonBadFrameSize: String = "expected 640 byte PCM frame"

enum ErrorStage(val wire: String):
  case WebsocketReceive extends ErrorStage("websocket_receive")
  case BatchFlush extends ErrorStage("batch_flush")
  case InferenceRequest extends ErrorStage("inference_request")
  case InferenceResponseParse extends ErrorStage("inference_response_parse")
  case WebsocketSend extends ErrorStage("websocket_send")

object ErrorStage:
  given Encoder[ErrorStage] = Encoder[String].contramap(_.wire)

enum ErrorKind(val wire: String):
  case Timeout extends ErrorKind("timeout")
  case PoolTimeout extends ErrorKind("pool_timeout")
  case Http5xx extends ErrorKind("http_5xx")
  case Http429 extends ErrorKind("http_429")
  case ConnectionReset extends ErrorKind("connection_reset")
  case ParseError extends ErrorKind("parse_error")
  case SendError extends ErrorKind("send_error")

object ErrorKind:
  given Encoder[ErrorKind] = Encoder[String].contramap(_.wire)

/** Strict snake_case decoder configuration shared by inbound wire types. */
private object DecoderConfig:
  given Configuration = Configuration.default.withSnakeCaseMemberNames.withStrictDecoding

/** Strict snake_case encoder configuration shared by outbound wire types. */
private object EncoderConfig:
  given Configuration = Configuration.default.withSnakeCaseMemberNames

/** First message the client sends on a fresh WebSocket. */
final case class StartMessage(`type`: String)

object StartMessage:

  given Decoder[StartMessage] = Decoder.instance { c =>
    val expected = Set("type")
    c.keys match
      case Some(keys) if keys.toSet != expected =>
        Left(DecodingFailure(s"unknown fields: ${keys.toSet -- expected}", c.history))
      case _ =>
        c.downField("type").as[String].flatMap { tpe =>
          if tpe == "start" then Right(StartMessage(tpe))
          else Left(DecodingFailure(s"expected type=start, got $tpe", c.history))
        }
  }

/** Inference server response. Decoded strictly: unknown fields are rejected. */
final case class InferResponse(
    rms: Double,
    zeroCrossings: Long,
    checksum: Long,
    samples: Long,
    transcript: String,
    audioBytes: Long
)

object InferResponse:
  import EncoderConfig.given

  private val expectedFields = Set("rms", "zero_crossings", "checksum", "samples", "transcript", "audio_bytes")

  given Decoder[InferResponse] = Decoder.instance { c =>
    c.keys match
      case Some(keys) if keys.toSet != expectedFields =>
        Left(DecodingFailure(s"unknown fields: ${keys.toSet -- expectedFields}", c.history))
      case _ =>
        for
          rms <- c.downField("rms").as[Double]
          zeroCrossings <- c.downField("zero_crossings").as[Long]
          checksum <- c.downField("checksum").as[Long]
          samples <- c.downField("samples").as[Long]
          transcript <- c.downField("transcript").as[String]
          audioBytes <- c.downField("audio_bytes").as[Long]
        yield InferResponse(rms, zeroCrossings, checksum, samples, transcript, audioBytes)
  }

  given Encoder[InferResponse] = ConfiguredEncoder.derived[InferResponse]

/** `partial` message sent back on each successful flush. Field order is wire-compatible with the Rust/Go/Java gateways
  * via {@link ConfiguredEncoder} and the declared case-class order.
  */
final case class PartialMessage(
    `type`: String,
    oldestFrameSeq: Long,
    newestFrameSeq: Long,
    frames: Int,
    rms: Double,
    zeroCrossings: Long,
    checksum: Long,
    samples: Long,
    transcript: String,
    audioBytes: Long,
    cpuPasses: Int,
    modelDelayMs: Long,
    flushLatenessMs: Double,
    inflightModelJobs: Int
)

object PartialMessage:
  import EncoderConfig.given
  given Encoder[PartialMessage] = ConfiguredEncoder.derived[PartialMessage]

  def of(
      infer: InferResponse,
      oldestSeq: Long,
      newestSeq: Long,
      frames: Int,
      cpuPasses: Int,
      modelDelayMs: Long,
      flushLatenessMs: Double
  ): PartialMessage = PartialMessage(
    `type` = Protocol.PartialType,
    oldestFrameSeq = oldestSeq,
    newestFrameSeq = newestSeq,
    frames = frames,
    rms = infer.rms,
    zeroCrossings = infer.zeroCrossings,
    checksum = infer.checksum,
    samples = infer.samples,
    transcript = infer.transcript,
    audioBytes = infer.audioBytes,
    cpuPasses = cpuPasses,
    modelDelayMs = modelDelayMs,
    flushLatenessMs = flushLatenessMs,
    inflightModelJobs = 0
  )

/** `error` message emitted when the gateway fails to produce a partial for a batch. */
final case class ErrorMessage(
    `type`: String,
    stage: ErrorStage,
    kind: ErrorKind,
    message: String,
    oldestFrameSeq: Long,
    newestFrameSeq: Long,
    frames: Int,
    audioBytes: Long,
    oldestAgeMs: Double,
    newestAgeMs: Double,
    flushLatenessMs: Double,
    inferenceElapsedMs: Option[Double],
    inflightGatewayBatches: Int,
    gatewayBufferFrames: Int,
    inferenceStatus: Option[Int],
    retryable: Boolean
)

object ErrorMessage:
  given Configuration = Configuration.default.withSnakeCaseMemberNames

  // ConfiguredEncoder drops None-valued fields by default; we inject JSON null for the two
  // Optional fields so the wire shape matches Rust/Go/Java (which serialize null explicitly).
  given Encoder[ErrorMessage] = ConfiguredEncoder.derived[ErrorMessage].mapJson { json =>
    List("inference_elapsed_ms", "inference_status").foldLeft(json) { (acc, key) =>
      if acc.hcursor.downField(key).succeeded then acc
      else acc.deepMerge(Json.obj(key -> Json.Null))
    }
  }

/** Sealed dispatch shape for the actor's outbox. */
enum OutboundMessage:
  case Partial(message: PartialMessage)
  case Error(message: ErrorMessage)

/** Internal per-frame state captured at WS receive time. Not on the wire. {@code receivedAtNanos} is a
  * [[System.nanoTime]] reading — relative elapsed-millis math only, never an absolute time.
  */
final case class Frame(seq: Long, payload: Array[Byte], receivedAtNanos: Long)

File: services/typescript-bun/src/protocol.ts
import * as v from "valibot";

import { FRAME_BYTES } from "./config.ts";

export { FRAME_BYTES };

export const CLOSE_PROTOCOL_ERROR = 1002;
export const CLOSE_UNSUPPORTED_DATA = 1003;
export const REASON_NEED_START = "first message must be start";
export const REASON_TEXT_AFTER_START = "expected binary PCM frames after start";
export const REASON_BAD_FRAME_SIZE = "expected 640 byte PCM frame";

export type ErrorStage =
  | "websocket_receive"
  | "batch_flush"
  | "inference_request"
  | "inference_response_parse"
  | "websocket_send";

export type ErrorKind =
  | "timeout"
  | "pool_timeout"
  | "http_5xx"
  | "http_429"
  | "connection_reset"
  | "parse_error"
  | "send_error";

const NonNegativeSafeIntegerSchema = v.pipe(v.number(), v.safeInteger(), v.minValue(0));
const PositiveSafeIntegerSchema = v.pipe(v.number(), v.safeInteger(), v.minValue(1));
const FiniteNumberSchema = v.pipe(v.number(), v.finite());

const StartMessageSchema = v.strictObject({
  type: v.literal("start"),
});

export const PredictionSchema = v.strictObject({
  rms: FiniteNumberSchema,
  zero_crossings: NonNegativeSafeIntegerSchema,
  checksum: NonNegativeSafeIntegerSchema,
  samples: PositiveSafeIntegerSchema,
  transcript: v.string(),
  audio_bytes: PositiveSafeIntegerSchema,
});

export type Prediction = v.InferOutput<typeof PredictionSchema>;

export type ParseResult<T> =
  | { readonly success: true; readonly output: T }
  | { readonly success: false; readonly message: string };

export interface Frame {
  readonly seq: number;
  readonly payload: Uint8Array;
  readonly receivedAtMs: number;
}

export interface PartialBuildInput {
  readonly oldestFrameSeq: number;
  readonly newestFrameSeq: number;
  readonly frames: number;
  readonly prediction: Prediction;
  readonly cpuPasses: number;
  readonly modelDelayMs: number;
  readonly flushLatenessMs: number;
}

export interface ErrorBuildInput {
  readonly stage: ErrorStage;
  readonly kind: ErrorKind;
  readonly message: string;
  readonly oldestFrameSeq: number;
  readonly newestFrameSeq: number;
  readonly frames: number;
  readonly audioBytes: number;
  readonly oldestAgeMs: number;
  readonly newestAgeMs: number;
  readonly flushLatenessMs: number;
  readonly inferenceElapsedMs: number | null;
  readonly inflightGatewayBatches: number;
  readonly gatewayBufferFrames: number;
  readonly inferenceStatus: number | null;
  readonly retryable: boolean;
}

export function parseStartMessage(text: string): ParseResult<{ type: "start" }> {
  let payload: unknown;
  try {
    payload = JSON.parse(text);
  } catch (error) {
    return { success: false, message: errorMessage(error) };
  }

  const result = v.safeParse(StartMessageSchema, payload);
  if (!result.success) {
    return { success: false, message: result.issues.map((issue) => issue.message).join("; ") };
  }
  return { success: true, output: result.output };
}

export function parsePrediction(payload: unknown): Prediction {
  return v.parse(PredictionSchema, payload);
}

export function isPcmFrame(message: unknown): boolean {
  return toBytes(message)?.byteLength === FRAME_BYTES;
}

export function toBytes(message: unknown): Uint8Array | null {
  if (typeof message === "string") {
    return null;
  }
  if (message instanceof Uint8Array) {
    return message;
  }
  if (message instanceof ArrayBuffer) {
    return new Uint8Array(message);
  }
  return null;
}

export function buildPartialMessage(input: PartialBuildInput): string {
  return JSON.stringify({
    type: "partial",
    oldest_frame_seq: input.oldestFrameSeq,
    newest_frame_seq: input.newestFrameSeq,
    frames: input.frames,
    rms: input.prediction.rms,
    zero_crossings: input.prediction.zero_crossings,
    checksum: input.prediction.checksum,
    samples: input.prediction.samples,
    transcript: input.prediction.transcript,
    audio_bytes: input.prediction.audio_bytes,
    cpu_passes: input.cpuPasses,
    model_delay_ms: input.modelDelayMs,
    flush_lateness_ms: roundMillis(input.flushLatenessMs),
    inflight_model_jobs: 0,
  });
}

export function buildErrorMessage(input: ErrorBuildInput): string {
  return JSON.stringify({
    type: "error",
    stage: input.stage,
    kind: input.kind,
    message: input.message,
    oldest_frame_seq: input.oldestFrameSeq,
    newest_frame_seq: input.newestFrameSeq,
    frames: input.frames,
    audio_bytes: input.audioBytes,
    oldest_age_ms: roundMillis(input.oldestAgeMs),
    newest_age_ms: roundMillis(input.newestAgeMs),
    flush_lateness_ms: roundMillis(input.flushLatenessMs),
    inference_elapsed_ms:
      input.inferenceElapsedMs === null ? null : roundMillis(input.inferenceElapsedMs),
    inflight_gateway_batches: input.inflightGatewayBatches,
    gateway_buffer_frames: input.gatewayBufferFrames,
    inference_status: input.inferenceStatus,
    retryable: input.retryable,
  });
}

export function concatFrames(frames: Frame[]): Uint8Array {
  const totalBytes = frames.reduce((total, frame) => total + frame.payload.byteLength, 0);
  const body = new Uint8Array(totalBytes);
  let offset = 0;
  for (const frame of frames) {
    body.set(frame.payload, offset);
    offset += frame.payload.byteLength;
  }
  return body;
}

export function roundMillis(value: number): number {
  return Math.round(value * 1000) / 1000;
}

export function errorMessage(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}
